Tests fail on CI due to Terraform not being installed

These fixtures need a `terraform:` block so that astro downloads Terraform.

We should probably also cache the .tvm directory.

```
--- FAIL: TestNoVariables (0.00s)
	Error Trace:	flags_test.go:29
	Error:      	Received unexpected error:
	            	failed to load YAML from file: fixtures/flags/no_variables.yaml; exec: "terraform": executable file not found in $PATH
	Test:       	TestNoVariables
--- FAIL: TestSimpleVariables (0.00s)
	Error Trace:	flags_test.go:39
	Error:      	Received unexpected error:
	            	failed to load YAML from file: fixtures/flags/simple_variables.yaml; exec: "terraform": executable file not found in $PATH
	Test:       	TestSimpleVariables
--- FAIL: TestMergeValues (0.00s)
	Error Trace:	flags_test.go:70
	Error:      	Received unexpected error:
	            	failed to load YAML from file: fixtures/flags/merge_values.yaml; exec: "terraform": executable file not found in $PATH
	Test:       	TestMergeValues
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,6 +2,7 @@ language: go
 before_install:
 - curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
 cache:
+- /home/travis/.tvm
 - vendor
 go:
  - "1.11.x"

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,8 +2,9 @@ language: go
 before_install:
 - curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
 cache:
-- /home/travis/.tvm
-- vendor
+  directories:
+    - /home/travis/.tvm
+    - vendor
 go:
  - "1.11.x"
 install:

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,12 @@
+language: go
+before_install:
+- curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
+cache:
+- vendor
+go:
+ - "1.11.x"
+install:
+- make vendor
+script:
+- make lint
+- make test

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,8 @@
 
 * Add Travis configuration, `make lint` and git precommit hook
 * Fix issue with `make test` always returning true even when tests fail
+* Fix a race condition that could cause failures due to astro downloading the
+  same version of Terraform twice
 
 ## 0.4.1 (October 3, 2018)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,8 +3,9 @@
 ## 0.4.2 (UNRELEASED, 2018)
 
 * Add Travis configuration, `make lint` and git precommit hook
+* Fix issue with make not recompiling when source files changed
 * Fix issue with `make test` always returning true even when tests fail
-* Fix a race condition that could cause failures due to astro downloading the
+* Fix race condition that could cause failures due to astro downloading the
   same version of Terraform twice
 
 ## 0.4.1 (October 3, 2018)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,10 @@
+# astro changelog
+
+## 0.4.2 (UNRELEASED, 2018)
+
+* Add Travis configuration, `make lint` and git precommit hook
+* Fix issue with `make test` always returning true even when tests fail
+
 ## 0.4.1 (October 3, 2018)
 
 * Output policy changes in unified diff format (#2)

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 SHELL = /bin/bash -o pipefail
 
-export GO111MODULE=on
+SRC = $(shell find . -name '*.go')
 
 define PRE_COMMIT_HOOK
 #!/bin/sh -xe
@@ -24,10 +24,10 @@ export PRE_COMMIT_HOOK
 bin:
 	mkdir -p bin
 
-bin/astro: bin vendor
+bin/astro: bin vendor $(SRC)
 	go build -o bin/astro github.com/uber/astro/astro/cli/astro
 
-bin/tvm: bin vendor
+bin/tvm: bin vendor $(SRC)
 	go build -o bin/tvm github.com/uber/astro/astro/tvm/cli/tvm
 
 .PHONY: clean

--- a/Makefile
+++ b/Makefile
@@ -1,5 +1,23 @@
+export GO111MODULE=on
+
+define PRE_COMMIT_HOOK
+#!/bin/sh -xe
+make lint
+endef
+
+install_hook := $(shell \
+    if [ -z "$$INSTALL_HOOK" ]; then \
+        INSTALL_HOOK=1 make .git/hooks/pre-commit; \
+    fi; \
+)
+
 .PHONY: all
 all: bin/astro bin/tvm
+
+export PRE_COMMIT_HOOK
+.git/hooks/pre-commit: Makefile
+	echo "$$PRE_COMMIT_HOOK" >.git/hooks/pre-commit
+	chmod +x .git/hooks/pre-commit
 
 bin:
 	mkdir -p bin
@@ -19,6 +37,13 @@ clean:
 install:
 	go install github.com/uber/astro/astro/cli/astro
 	go install github.com/uber/astro/astro/tvm/cli/tvm
+
+.PHONY: lint
+lint:
+	@f="$$(find . -name '*.go' ! -path './vendor/*' | xargs grep -L 'Licensed under the Apache License')" || { \
+		echo "ERROR: Files missing license header:"$$'\n'"$$f" >&2; \
+		exit 1; \
+	};
 
 .PHONY: test
 test: vendor

--- a/Makefile
+++ b/Makefile
@@ -53,5 +53,6 @@ test: vendor
 	go test -timeout 1m -coverprofile=.coverage.out ./... \
 		|grep -v -E '^\?'
 
+.PHONY: vendor
 vendor:
 	dep ensure

--- a/Makefile
+++ b/Makefile
@@ -1,3 +1,5 @@
+SHELL = /bin/bash -o pipefail
+
 export GO111MODULE=on
 
 define PRE_COMMIT_HOOK

--- a/Makefile
+++ b/Makefile
@@ -42,10 +42,11 @@ install:
 
 .PHONY: lint
 lint:
-	@f="$$(find . -name '*.go' ! -path './vendor/*' | xargs grep -L 'Licensed under the Apache License')" || { \
+	@f="$$(find . -name '*.go' ! -path './vendor/*' | xargs grep -L 'Licensed under the Apache License')"; \
+	if [ ! -z "$$f" ]; then \
 		echo "ERROR: Files missing license header:"$$'\n'"$$f" >&2; \
 		exit 1; \
-	};
+	fi;
 
 .PHONY: test
 test: vendor

--- a/astro/cli/astro/cmd/cmd_test.go
+++ b/astro/cli/astro/cmd/cmd_test.go
@@ -90,7 +90,10 @@ func TestMain(m *testing.M) {
 	// Download Terraform versions first so that multiple tests don't
 	// try to do it in parallel.
 	for _, version := range terraformVersionsToTest {
-		terraformVersionRepo.Get(version)
+		if _, err := terraformVersionRepo.Get(version); err != nil {
+			fmt.Fprint(os.Stderr, err)
+			os.Exit(1)
+		}
 	}
 
 	os.Exit(m.Run())

--- a/astro/cli/astro/cmd/fixtures/flags/merge_values.yaml
+++ b/astro/cli/astro/cmd/fixtures/flags/merge_values.yaml
@@ -1,5 +1,8 @@
 ---
 
+terraform:
+  version: 0.11.7
+
 modules:
   - name: foo_mgmt
     path: .

--- a/astro/cli/astro/cmd/fixtures/flags/no_variables.yaml
+++ b/astro/cli/astro/cmd/fixtures/flags/no_variables.yaml
@@ -1,5 +1,8 @@
 ---
 
+terraform:
+  version: 0.11.7
+
 modules:
   - name: foo
     path: .

--- a/astro/cli/astro/cmd/fixtures/flags/simple_variables.yaml
+++ b/astro/cli/astro/cmd/fixtures/flags/simple_variables.yaml
@@ -1,5 +1,8 @@
 ---
 
+terraform:
+  version: 0.11.7
+
 modules:
   - name: foo
     path: .

--- a/astro/tvm/versionrepo.go
+++ b/astro/tvm/versionrepo.go
@@ -47,7 +47,13 @@ type VersionRepo struct {
 	arch     string
 	platform string
 
-	lock *sync.Map
+	// locks is a map of mutexes. There is one mutex created on demand for
+	// every Terraform version requested from tvm. The mutex prevents tvm from
+	// downloading the same version of Terraform multiple times. If multiple
+	// threads request the same version of Terraform, only one of them will
+	// trigger the download and the rest will block until the download is
+	// complete.
+	locks *sync.Map
 }
 
 // NewVersionRepo creates a new VersionRepo. The arch will
@@ -68,7 +74,7 @@ func NewVersionRepo(repoPath string, arch string, platform string) (*VersionRepo
 	}
 
 	return &VersionRepo{
-		lock:     &sync.Map{},
+		locks:    &sync.Map{},
 		repoPath: repoPath,
 		arch:     arch,
 		platform: platform,
@@ -141,8 +147,11 @@ func (r *VersionRepo) exists(version string) bool {
 	return utils.FileExists(r.terraformPath(version))
 }
 
+// getLock returns a mutex for the specified Terraform version which is used to
+// prevent multiple threads from downloading the same version of Terraform at
+// the same time.
 func (r *VersionRepo) getLock(version string) *sync.Mutex {
-	v, _ := r.lock.LoadOrStore(version, &sync.Mutex{})
+	v, _ := r.locks.LoadOrStore(version, &sync.Mutex{})
 	return v.(*sync.Mutex)
 }
 
@@ -152,6 +161,8 @@ func (r *VersionRepo) getLock(version string) *sync.Mutex {
 func (r *VersionRepo) Get(version string) (string, error) {
 	lock := r.getLock(version)
 
+	// Lock() here will block if another thread is currently downloading
+	// Terraform.
 	lock.Lock()
 	defer lock.Unlock()
 

--- a/astro/tvm/versionrepo.go
+++ b/astro/tvm/versionrepo.go
@@ -47,7 +47,7 @@ type VersionRepo struct {
 	arch     string
 	platform string
 
-	downloadLock *sync.Map
+	lock *sync.Map
 }
 
 // NewVersionRepo creates a new VersionRepo. The arch will
@@ -62,16 +62,16 @@ func NewVersionRepo(repoPath string, arch string, platform string) (*VersionRepo
 		repoPath = filepath.Join(home, ".tvm")
 	}
 
-	if !utils.FileExists(repoPath) {
-		if err := os.Mkdir(repoPath, 0755); err != nil {
-			return nil, err
-		}
+	// Create directory if it doesn't exist
+	if err := os.Mkdir(repoPath, 0755); err != nil && !os.IsExist(err) {
+		return nil, err
 	}
+
 	return &VersionRepo{
-		downloadLock: &sync.Map{},
-		repoPath:     repoPath,
-		arch:         arch,
-		platform:     platform,
+		lock:     &sync.Map{},
+		repoPath: repoPath,
+		arch:     arch,
+		platform: platform,
 	}, nil
 }
 
@@ -92,11 +92,6 @@ func (r *VersionRepo) dir(version string) string {
 // returns the path to the downloaded file or an error if there was a
 // problem.
 func (r *VersionRepo) download(version string) (string, error) {
-	downloadLock := r.getLock(version)
-
-	downloadLock.Lock()
-	defer downloadLock.Unlock()
-
 	url := fmt.Sprintf(terraformZipFileDownloadURL, version, version, r.platform, r.arch)
 
 	// Temporary directory for downloading Terraform and extracting the zip file
@@ -128,7 +123,9 @@ func (r *VersionRepo) download(version string) (string, error) {
 	targetDir := r.dir(version)
 
 	// Make repo dir
-	os.MkdirAll(targetDir, os.ModePerm)
+	if err := os.MkdirAll(targetDir, os.ModePerm); err != nil {
+		return "", err
+	}
 
 	// Move binary to repo path
 	if err := os.Rename(terraformBinaryPath, path.Join(targetDir, "terraform")); err != nil {
@@ -145,7 +142,7 @@ func (r *VersionRepo) exists(version string) bool {
 }
 
 func (r *VersionRepo) getLock(version string) *sync.Mutex {
-	v, _ := r.downloadLock.LoadOrStore(version, &sync.Mutex{})
+	v, _ := r.lock.LoadOrStore(version, &sync.Mutex{})
 	return v.(*sync.Mutex)
 }
 
@@ -153,6 +150,11 @@ func (r *VersionRepo) getLock(version string) *sync.Mutex {
 // that version. If the binary doesn't exist, it will be downloaded from
 // the Terraform website automatically.
 func (r *VersionRepo) Get(version string) (string, error) {
+	lock := r.getLock(version)
+
+	lock.Lock()
+	defer lock.Unlock()
+
 	path := r.terraformPath(version)
 	if !utils.FileExists(path) {
 		return r.download(version)

--- a/astro/tvm/versionrepo.go
+++ b/astro/tvm/versionrepo.go
@@ -161,8 +161,8 @@ func (r *VersionRepo) getLock(version string) *sync.Mutex {
 func (r *VersionRepo) Get(version string) (string, error) {
 	lock := r.getLock(version)
 
-	// Lock() here will block if another thread is currently downloading
-	// Terraform.
+	// Lock() here will block and wait if another thread is currently
+	// downloading Terraform.
 	lock.Lock()
 	defer lock.Unlock()
 

--- a/astro/tvm/versionrepo.go
+++ b/astro/tvm/versionrepo.go
@@ -23,6 +23,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"path"
 	"path/filepath"
 	"runtime"
 	"sync"
@@ -98,28 +99,40 @@ func (r *VersionRepo) download(version string) (string, error) {
 
 	url := fmt.Sprintf(terraformZipFileDownloadURL, version, version, r.platform, r.arch)
 
-	// Temporary file for zip file
-	tmpfile, err := ioutil.TempFile("", "terraform")
+	// Temporary directory for downloading Terraform and extracting the zip file
+	tmpDir, err := ioutil.TempDir("", "terraform")
 	if err != nil {
 		return "", err
 	}
-	defer os.Remove(tmpfile.Name())
+	defer os.RemoveAll(tmpDir)
 
-	// Download zip file
-	if err := downloadFile(url, tmpfile.Name()); err != nil {
+	zipFilePath := path.Join(tmpDir, "terraform.zip")
+
+	// Download Terraform zip file
+	if err := downloadFile(url, zipFilePath); err != nil {
 		return "", err
+	}
+
+	// Extract contents of zip file
+	if err := unzip(zipFilePath, tmpDir); err != nil {
+		return "", err
+	}
+
+	terraformBinaryPath := path.Join(tmpDir, "terraform")
+
+	// Check the binary is there
+	if !utils.FileExists(terraformBinaryPath) {
+		return "", errors.New("Terraform binary missing from zip file")
 	}
 
 	targetDir := r.dir(version)
 
-	// Extract contents of zip file to repo path
-	if err := unzip(tmpfile.Name(), targetDir); err != nil {
-		return "", err
-	}
+	// Make repo dir
+	os.MkdirAll(targetDir, os.ModePerm)
 
-	// Check the binary is there
-	if !r.exists(version) {
-		return "", errors.New("Terraform binary missing from zip file")
+	// Move binary to repo path
+	if err := os.Rename(terraformBinaryPath, path.Join(targetDir, "terraform")); err != nil {
+		return "", err
 	}
 
 	return r.terraformPath(version), nil

--- a/astro/tvm/versionrepo_test.go
+++ b/astro/tvm/versionrepo_test.go
@@ -27,7 +27,41 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-func TestDownload(t *testing.T) {
+// TestDownloadSameRepo does multiple downloads in parallel in the same repo
+// which should test the syncMap / download lock.
+func TestDownloadSameRepo(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping download test in short mode.")
+	}
+
+	tmpdir, err := ioutil.TempDir("", "terraform-tests")
+	require.NoError(t, err)
+
+	defer os.RemoveAll(tmpdir)
+
+	for i := 1; i <= 3; i++ {
+		t.Run(string(i), func(t *testing.T) {
+			t.Parallel()
+
+			versions, err := tvm.NewVersionRepoForCurrentSystem(tmpdir)
+			require.NoError(t, err)
+
+			terraformBinary, err := versions.Get("0.7.13")
+			require.NoError(t, err)
+
+			version, err := tvm.InspectVersion(terraformBinary)
+			require.NoError(t, err)
+
+			assert.Equal(t, "0.7.13", version.String())
+		})
+	}
+}
+
+// TestDownloadDifferentRepo does multiple downloads in parallel
+// reinitializing the repo every time. This will cause multiple downloads to
+// occur but it will test for races that can occur when executing a binary that
+// is currently being written to.
+func TestDownloadDifferentRepo(t *testing.T) {
 	if testing.Short() {
 		t.Skip("skipping download test in short mode.")
 	}
@@ -40,11 +74,17 @@ func TestDownload(t *testing.T) {
 	versions, err := tvm.NewVersionRepoForCurrentSystem(tmpdir)
 	require.NoError(t, err)
 
-	terraformBinary, err := versions.Get("0.7.13")
-	require.NoError(t, err)
+	for i := 1; i <= 3; i++ {
+		t.Run(string(i), func(t *testing.T) {
+			t.Parallel()
 
-	version, err := tvm.InspectVersion(terraformBinary)
-	require.NoError(t, err)
+			terraformBinary, err := versions.Get("0.7.13")
+			require.NoError(t, err)
 
-	assert.Equal(t, "0.7.13", version.String())
+			version, err := tvm.InspectVersion(terraformBinary)
+			require.NoError(t, err)
+
+			assert.Equal(t, "0.7.13", version.String())
+		})
+	}
 }

